File: code/jvm/service/src/main/java/pt/isel/services/channels/ChannelService.java
package pt.isel.services.channels;

import jakarta.inject.Named;
import pt.isel.domain.channels.AccessType;
import pt.isel.domain.channels.Channel;
import pt.isel.domain.channels.ChannelMember;
import pt.isel.domain.common.ChannelError;
import pt.isel.domain.common.Either;
import pt.isel.domain.common.MessageError;
import pt.isel.domain.invitations.InvitationStatus;
import pt.isel.domain.users.User;
import pt.isel.domain.users.UserInfo;
import pt.isel.repositories.Transaction;
import pt.isel.repositories.TransactionManager;

import java.time.Clock;
import java.time.LocalDateTime;
import java.util.List;

@Named
public class ChannelService {
    private final TransactionManager trxManager;
    private final Clock clock;

    public ChannelService(TransactionManager trxManager, Clock clock) {
        this.trxManager = trxManager;
        this.clock = clock;
    }

    public Either<ChannelError, Channel> createChannel(String name, Long ownerId, boolean isPublic) {
        if (name == null || name.isBlank()) return Either.failure(new ChannelError.EmptyChannelName());
        if (name.length() > 30) return Either.failure(new ChannelError.InvalidChannelNameLength());

        return trxManager.run(trx -> {
            User owner = trx.repoUsers().findById(ownerId);
            if (owner == null) return Either.failure(new ChannelError.UserNotFound());

            UserInfo ownerInfo = new UserInfo(owner.id(), owner.username());

            try {
                Channel newChannel = trx.repoChannels().create(name, ownerInfo, isPublic);
                trx.repoMemberships().addUserToChannel(ownerInfo, newChannel, AccessType.READ_WRITE);
                return Either.success(newChannel);
            } catch (Exception e) {
                if (e.getMessage() != null && e.getMessage().contains("channels_name_key")) {
                    return Either.failure(new ChannelError.ChannelAlreadyExists());
                }
                throw e;
            }
        });
    }

    public Either<ChannelError, Channel> getChannelById(Long channelId) {
        return trxManager.run(trx -> {
            Channel channel = trx.repoChannels().findById(channelId);
            return channel != null ? Either.success(channel) : Either.failure(new ChannelError.ChannelNotFound());
        });
    }

    public Either<ChannelError, String> deleteChannel(Long ownerId, Long channelId) {
        return trxManager.run(trx -> checkUserIsOwner(trx, ownerId, channelId)
                .flatMap(pair -> {
                    trx.repoChannels().deleteById(pair.channel().id());
                    return Either.success("Channel '" + pair.channel().name() + "' was deleted successfully.");
                }));
    }

    public Either<ChannelError, List<Channel>> getJoinedChannels(Long userId, int limit, int offset) {
        return trxManager.run(trx -> {
            if (trx.repoUsers().findById(userId) == null) return Either.failure(new ChannelError.UserNotFound());
            List<Channel> channels = trx.repoMemberships().findAllChannelsForUser(userId, limit, offset).stream().map(ChannelMember::channel).toList();
            return Either.success(channels);
        });
    }

    public Either<ChannelError, List<UserInfo>> getUsersInChannel(Long channelId, int limit, int offset) {
        return trxManager.run(trx -> {
            List<UserInfo> users = trx.repoMemberships().findAllMembersInChannel(channelId, limit, offset).stream().map(ChannelMember::user).toList();
            return Either.success(users);
        });
    }

    public Either<ChannelError, Channel> editChannel(Long ownerId, Long channelId, String name, boolean isPublic) {
        if (name == null || name.isBlank()) return Either.failure(new ChannelError.EmptyChannelName());
        if (name.length() > 30) return Either.failure(new ChannelError.InvalidChannelNameLength());

        return trxManager.run(trx -> checkUserIsOwner(trx, ownerId, channelId)
                .flatMap(pair -> {
                    Channel channel = pair.channel();
                    try {
                        Channel updatedChannel = new Channel(channel.id(), name, channel.owner(), isPublic);
                        trx.repoChannels().save(updatedChannel);
                        return Either.success(updatedChannel);
                    } catch (Exception e) {
                        if (e.getMessage() != null && e.getMessage().contains("channels_name_key")) {
                            return Either.failure(new ChannelError.ChannelAlreadyExists());
                        }
                        throw e;
                    }
                }));
    }

    public Either<MessageError, AccessType> getAccessType(Long requesterId, Long targetUserId, Long channelId) {
        return trxManager.run(trx -> {
            var targetMembership = trx.repoMemberships().findUserInChannel(targetUserId, channelId);
            if (targetMembership == null) return Either.failure(new MessageError.UserNotInChannel());

            if (!requesterId.equals(targetUserId)) {
                if (trx.repoMemberships().findUserInChannel(requesterId, channelId) == null) {
                    return Either.failure(new MessageError.UserNotAuthorized());
                }
            }
            return Either.success(targetMembership.accessType());
        });
    }

    public Either<ChannelError, ChannelMember> editMemberAccess(Long ownerId, Long channelId, Long userId, AccessType accessType) {
        return trxManager.run(trx -> checkUserCanEditMember(trx, ownerId, channelId, userId)
                .flatMap(triple -> {
                    var userMembership = triple.userMembership();
                    ChannelMember updatedMembership = new ChannelMember(userMembership.id(), userMembership.user(), userMembership.channel(), accessType);
                    trx.repoMemberships().save(updatedMembership);
                    return Either.success(updatedMembership);
                }));
    }

    public Either<ChannelError, List<Channel>> searchChannels(String query, int limit, int offset) {
        return trxManager.run(trx -> {
            List<Channel> channels = (query == null || query.isBlank())
                    ? trx.repoChannels().findAllPublicChannels(limit, offset)
                    : trx.repoChannels().searchByName(query, limit, offset);
            return Either.success(channels);
        });
    }

    public Either<ChannelError, String> joinPublicChannel(Long userId, Long channelId) {
        return trxManager.run(trx -> {
            User user = trx.repoUsers().findById(userId);
            if (user == null) return Either.failure(new ChannelError.UserNotFound());

            Channel channel = trx.repoChannels().findById(channelId);
            if (channel == null) return Either.failure(new ChannelError.ChannelNotFound());

            if (!channel.isPublic()) return Either.failure(new ChannelError.ChannelIsPrivate());

            if (trx.repoMemberships().findUserInChannel(user.id(), channel.id()) != null) {
                return Either.failure(new ChannelError.UserAlreadyInChannel());
            }

            trx.repoMemberships().addUserToChannel(new UserInfo(user.id(), user.username()), channel, AccessType.READ_WRITE);
            return Either.success("Joined public channel '" + channel.name() + "'.");
        });
    }

    public Either<ChannelError, String> joinPrivateChannel(Long userId, String token) {
        return trxManager.run(trx -> {
            User user = trx.repoUsers().findById(userId);
            if (user == null) return Either.failure(new ChannelError.UserNotFound());

            var invitation = trx.repoInvitations().findByToken(token);
            if (invitation == null) return Either.failure(new ChannelError.TokenNotFound());

            if (invitation.expiresAt().isBefore(LocalDateTime.now(clock)))
                return Either.failure(new ChannelError.InvitationExpired());
            if (invitation.status() != InvitationStatus.PENDING)
                return Either.failure(new ChannelError.InvitationAlreadyUsed());

            Channel channel = trx.repoChannels().findById(invitation.channel().id());
            if (channel == null) return Either.failure(new ChannelError.ChannelNotFound());

            if (trx.repoMemberships().findUserInChannel(user.id(), channel.id()) != null) {
                return Either.failure(new ChannelError.UserAlreadyInChannel());
            }

            if (!trx.repoInvitations().consume(invitation.id())) {
                return Either.failure(new ChannelError.InvitationAlreadyUsed());
            }

            trx.repoMemberships().addUserToChannel(new UserInfo(user.id(), user.username()), channel, invitation.accessType());
            return Either.success("Joined private channel '" + channel.name() + "'");
        });
    }

    public Either<ChannelError, String> leaveChannel(Long channelId, Long userId) {
        return trxManager.run(trx -> {
            Channel channel = trx.repoChannels().findById(channelId);
            if (channel == null) return Either.failure(new ChannelError.ChannelNotFound());

            ChannelMember membership = trx.repoMemberships().findUserInChannel(userId, channelId);
            if (membership == null) return Either.failure(new ChannelError.UserNotInChannel());

            if (channel.owner().id().equals(userId)) return Either.failure(new ChannelError.OwnerCannotLeave());

            trx.repoMemberships().deleteById(membership.id());
            return Either.success("Left channel '" + channel.name() + "'");
        });
    }

    private Either<ChannelError, UserChannelPair> checkUserIsOwner(Transaction trx, Long userId, Long channelId) {
        User user = trx.repoUsers().findById(userId);
        if (user == null) return Either.failure(new ChannelError.UserNotFound());

        Channel channel = trx.repoChannels().findById(channelId);
        if (channel == null) return Either.failure(new ChannelError.ChannelNotFound());

        if (!channel.owner().id().equals(user.id())) return Either.failure(new ChannelError.UserNotOwner());

        return Either.success(new UserChannelPair(user, channel));
    }

    private Either<ChannelError, MemberEditTriple> checkUserCanEditMember(Transaction trx, Long ownerId, Long channelId, Long userId) {
        User owner = trx.repoUsers().findById(ownerId);
        if (owner == null) return Either.failure(new ChannelError.UserNotFound());

        User userToEdit = trx.repoUsers().findById(userId);
        if (userToEdit == null) return Either.failure(new ChannelError.UserNotFound());

        Channel channel = trx.repoChannels().findById(channelId);
        if (channel == null) return Either.failure(new ChannelError.ChannelNotFound());

        ChannelMember ownerMembership = trx.repoMemberships().findUserInChannel(owner.id(), channel.id());
        if (ownerMembership == null) return Either.failure(new ChannelError.UserNotInChannel());

        ChannelMember userMembership = trx.repoMemberships().findUserInChannel(userToEdit.id(), channel.id());
        if (userMembership == null) return Either.failure(new ChannelError.UserNotInChannel());

        if (!channel.owner().id().equals(owner.id())) return Either.failure(new ChannelError.UserNotOwner());
        if (userToEdit.id().equals(channel.owner().id())) return Either.failure(new ChannelError.UserIsOwner());

        return Either.success(new MemberEditTriple(owner, userToEdit, userMembership));
    }

    private record UserChannelPair(User user, Channel channel) {
    }

    private record MemberEditTriple(User owner, User userToEdit, ChannelMember userMembership) {
    }
}

File: code/jvm/service/src/main/kotlin/pt/isel/services/ChannelService.kt
package pt.isel.services

import jakarta.inject.Named
import pt.isel.domain.channel.AccessType
import pt.isel.domain.channel.Channel
import pt.isel.domain.channel.ChannelMember
import pt.isel.domain.invitation.InvitationStatus
import pt.isel.domain.user.User
import pt.isel.domain.user.UserInfo
import pt.isel.repositories.Transaction
import pt.isel.repositories.TransactionManager
import java.time.LocalDateTime

@Named
class ChannelService(
    private val trxManager: TransactionManager,
) {
    fun createChannel(
        name: String,
        ownerId: Long,
        isPublic: Boolean,
    ): Either<ChannelError, Channel> {
        if (name.isBlank()) return failure(ChannelError.EmptyChannelName)
        if (name.length !in 1..30) return failure(ChannelError.InvalidChannelNameLength)

        return trxManager.run {
            val owner = repoUsers.findById(ownerId) ?: return@run failure(ChannelError.UserNotFound)

            if (repoChannels.findByName(name) != null) {
                return@run failure(ChannelError.ChannelAlreadyExists)
            }
            val ownerInfo = UserInfo(owner.id, owner.username)
            val newChannel = repoChannels.create(name, ownerInfo, isPublic)
            repoMemberships.addUserToChannel(ownerInfo, newChannel, AccessType.READ_WRITE)
            success(newChannel)
        }
    }

    fun getChannelById(channelId: Long): Either<ChannelError, Channel> =
        trxManager.run {
            repoChannels.findById(channelId)?.let { success(it) }
                ?: return@run failure(ChannelError.ChannelNotFound)
        }

    fun deleteChannel(
        ownerId: Long,
        channelId: Long,
    ): Either<ChannelError, String> =
        trxManager.run {
            when (val result = checkUserIsOwner(ownerId, channelId)) {
                is Failure -> {
                    result
                }

                is Success -> {
                    val (_, channel) = result.value
                    repoChannels.deleteById(channel.id)
                    success("Channel '${channel.name}' was deleted successfully.")
                }
            }
        }

    fun getJoinedChannels(
        userId: Long,
        limit: Int = 50,
        offset: Int = 0,
    ): Either<ChannelError, List<Channel>> {
        if (limit <= 0) return failure(ChannelError.InvalidLimit)
        if (offset < 0) return failure(ChannelError.InvalidOffset)

        return trxManager.run {
            if (repoUsers.findById(userId) == null) return@run failure(ChannelError.UserNotFound)
            val channels =
                repoMemberships.findAllChannelsForUser(userId, limit, offset).map { it.channel }
            success(channels)
        }
    }

    fun getUsersInChannel(
        channelId: Long,
        limit: Int = 50,
        offset: Int = 0,
    ): Either<ChannelError, List<UserInfo>> {
        if (limit <= 0) return failure(ChannelError.InvalidLimit)
        if (offset < 0) return failure(ChannelError.InvalidOffset)

        return trxManager.run {
            val users =
                repoMemberships.findAllMembersInChannel(channelId, limit, offset).map { it.user }
            success(users)
        }
    }

    fun editChannel(
        ownerId: Long,
        channelId: Long,
        name: String,
        isPublic: Boolean,
    ): Either<ChannelError, Channel> {
        if (name.isBlank()) return failure(ChannelError.EmptyChannelName)
        if (name.length !in 1..30) return failure(ChannelError.InvalidChannelNameLength)

        return trxManager.run {
            when (val result = checkUserIsOwner(ownerId, channelId)) {
                is Failure -> {
                    result
                }

                is Success -> {
                    val (_, channel) = result.value
                    val updatedChannel = channel.copy(name = name, isPublic = isPublic)
                    repoChannels.save(updatedChannel)
                    success(updatedChannel)
                }
            }
        }
    }

    fun getAccessType(
        userId: Long,
        channelId: Long,
    ): Either<MessageError, AccessType> {
        return trxManager.run {
            val membership =
                repoMemberships.findUserInChannel(userId, channelId)
                    ?: return@run failure(MessageError.UserNotInChannel)

            success(membership.accessType)
        }
    }

    fun editMemberAccess(
        ownerId: Long,
        channelId: Long,
        userId: Long,
        accessType: AccessType,
    ): Either<ChannelError, ChannelMember> =
        trxManager.run {
            when (val result = checkUserCanEditMember(ownerId, channelId, userId)) {
                is Failure -> {
                    result
                }

                is Success -> {
                    val (_, _, userMembership) = result.value
                    val updatedMembership = userMembership.copy(accessType = accessType)
                    repoMemberships.save(updatedMembership)
                    success(updatedMembership)
                }
            }
        }

    fun searchChannels(
        query: String,
        limit: Int = 50,
        offset: Int = 0,
    ): Either<ChannelError, List<Channel>> {
        if (limit <= 0) return failure(ChannelError.InvalidLimit)
        if (offset < 0) return failure(ChannelError.InvalidOffset)

        return trxManager.run {
            val channels =
                if (query.isBlank()) {
                    repoChannels.findAllPublicChannels(limit, offset)
                } else {
                    repoChannels.searchByName(query, limit, offset).filter {
                        it.name.contains(query, ignoreCase = true)
                    }
                }
            success(channels)
        }
    }

    fun joinPublicChannel(
        userId: Long,
        channelId: Long,
    ): Either<ChannelError, String> =
        trxManager.run {
            val user = repoUsers.findById(userId) ?: return@run failure(ChannelError.UserNotFound)
            val channel =
                repoChannels.findById(channelId) ?: return@run failure(ChannelError.ChannelNotFound)

            if (repoMemberships.findUserInChannel(user.id, channel.id) != null) {
                return@run failure(ChannelError.UserAlreadyInChannel)
            }
            val userInfo = UserInfo(user.id, user.username)
            repoMemberships.addUserToChannel(userInfo, channel, AccessType.READ_WRITE)
            return@run success("Joined public channel '${channel.name}'.")
        }

    fun leaveChannel(
        channelId: Long,
        userId: Long,
    ): Either<ChannelError, String> {
        return trxManager.run {
            val channel =
                repoChannels.findById(channelId) ?: return@run failure(ChannelError.ChannelNotFound)
            val membership =
                repoMemberships.findUserInChannel(userId, channelId)
                    ?: return@run failure(ChannelError.UserNotInChannel)

            if (channel.owner.id == userId) {
                return@run failure(ChannelError.OwnerCannotLeave)
            }

            repoMemberships.deleteById(membership.id)
            success("Left channel '${channel.name}'")
        }
    }

    fun joinPrivateChannel(
        userId: Long,
        token: String,
    ): Either<ChannelError, String> =
        trxManager.run {
            val user = repoUsers.findById(userId) ?: return@run failure(ChannelError.UserNotFound)
            val invitation =
                repoInvitations.findByToken(token)
                    ?: return@run failure(ChannelError.TokenNotFound)

            if (invitation.expiresAt.isBefore(LocalDateTime.now())) {
                return@run failure(ChannelError.InvitationExpired)
            }
            if (invitation.status != InvitationStatus.PENDING) {
                return@run failure(ChannelError.InvitationAlreadyUsed)
            }

            val channel =
                repoChannels.findById(invitation.channel.id)
                    ?: return@run failure(ChannelError.ChannelNotFound)

            if (repoMemberships.findUserInChannel(user.id, channel.id) != null) {
                return@run failure(ChannelError.UserAlreadyInChannel)
            }
            val userInfo = UserInfo(user.id, user.username)
            repoMemberships.addUserToChannel(userInfo, channel, invitation.accessType)
            repoInvitations.save(invitation.copy(status = InvitationStatus.ACCEPTED))
            success("Joined private channel '${channel.name}'")
        }

    private fun Transaction.checkUserIsOwner(
        userId: Long,
        channelId: Long,
    ): Either<ChannelError, Pair<User, Channel>> {
        val user = repoUsers.findById(userId) ?: return failure(ChannelError.UserNotFound)
        val channel =
            repoChannels.findById(channelId) ?: return failure(ChannelError.ChannelNotFound)
        if (channel.owner.id != user.id) {
            return failure(ChannelError.UserNotOwner)
        }
        return success(user to channel)
    }

    private fun Transaction.checkUserCanEditMember(
        ownerId: Long,
        channelId: Long,
        userId: Long,
    ): Either<ChannelError, Triple<User, User, ChannelMember>> {
        val owner = repoUsers.findById(ownerId) ?: return failure(ChannelError.UserNotFound)
        val userToEdit = repoUsers.findById(userId) ?: return failure(ChannelError.UserNotFound)
        val channel =
            repoChannels.findById(channelId) ?: return failure(ChannelError.ChannelNotFound)

        val ownerMembership =
            repoMemberships.findUserInChannel(owner.id, channel.id)
                ?: return failure(ChannelError.UserNotInChannel)
        val userMembership =
            repoMemberships.findUserInChannel(userToEdit.id, channel.id)
                ?: return failure(ChannelError.UserNotInChannel)

        if (channel.owner.id != owner.id) return failure(ChannelError.UserNotOwner)
        if (userToEdit.id == channel.owner.id) return failure(ChannelError.UserIsOwner)
        if (ownerMembership.accessType == AccessType.READ_ONLY) return failure(ChannelError.UserNotAuthorized)

        return success(Triple(owner, userToEdit, userMembership))
    }
}
